src/NumericalAlgorithms/LinearOperators/IndefiniteIntegral.cpp

```diff
@@ -0,0 +1,151 @@
+#include "Domain/Mesh.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+template <size_t Dim>
+void indefinite_integral_impl(const gsl::not_null<DataVector*> integral,
+                              const DataVector& integrand,
+                              const Mesh<Dim>& mesh,
+                              const size_t dim) noexcept {
+  if (UNLIKELY(integral->size() != integrand.size())) {
+    *integral = DataVector(integrand.size());
+  }
+  const Mesh<1> mesh_1d{mesh.extents(dim), gsl::at(mesh.basis(), dim),
+                        gsl::at(mesh.quadrature(), dim)};
+  const size_t num_pts = mesh_1d.number_of_grid_points();
+  const Matrix& indefenite_integration_matrix =
+      Spectral::integration_matrix(mesh_1d);
+  for (StripeIterator stripe_it(mesh.extents(), dim); stripe_it; ++stripe_it) {
+    dgemv_('N', num_pts, num_pts, 1.0, indefenite_integration_matrix.data(),
+           num_pts, integrand.data() + stripe_it.offset(),  // NOLINT
+           stripe_it.stride(), 0.0,
+           integral->data() + stripe_it.offset(),  // NOLINT
+           stripe_it.stride());
+  }
+}
+}  // namespace
+
+template <size_t Dim>
+void indefinite_integral(const gsl::not_null<DataVector*> integral,
+                         const DataVector& integrand,
+                         const Mesh<Dim>& mesh) noexcept {
+  if (UNLIKELY(integral->size() != integrand.size())) {
+    *integral = DataVector(integrand.size());
+  }
+  if (Dim == 1) {
+    indefinite_integral_impl(integral, integrand, mesh, 0);
+  } else if (Dim == 2) {
+    DataVector temp(integral->size());
+    indefinite_integral_impl(&temp, integrand, mesh, 0);
+    indefinite_integral_impl(integral, temp, mesh, 1);
+  } else if (Dim == 3) {
+    DataVector temp(integral->size());
+    indefinite_integral_impl(integral, integrand, mesh, 0);
+    indefinite_integral_impl(&temp, *integral, mesh, 1);
+    indefinite_integral_impl(integral, temp, mesh, 2);
+  } else {
+    ERROR("Can only do 1, 2, or 3 dimensional indefinite integrals");
+  }
+}
+
+template <size_t Dim>
+DataVector indefinite_integral(const DataVector& integrand,
+                               const Mesh<Dim>& mesh) noexcept {
+  DataVector integral(integrand.size());
+  indefinite_integral(&integral, integrand, mesh);
+  return integral;
+}
+
+template <size_t Dim>
+void indefinite_integral(const gsl::not_null<DataVector*> integral,
+                         const DataVector& integrand, const Mesh<Dim>& mesh,
+                         const size_t dim_to_integrate) noexcept {
+  if (UNLIKELY(integral->size() != integrand.size())) {
+    *integral = DataVector(integrand.size());
+  }
+  indefinite_integral_impl(integral, integrand, mesh, dim_to_integrate);
+}
+
+template <size_t Dim>
+DataVector indefinite_integral(const DataVector& integrand,
```

Comment:
Looks like I can't both specify boundary conditions and an axis I want to integrate, which would be the use-case I would want. Not that I can't add it myself later :smile:.